test/e2e/subroutes_test.go

```diff
@@ -109,66 +236,157 @@ func TestSubrouteVisibilityChange(t *testing.T) {
+	resources, err := v1a1test.CreateRunLatestServiceReady(t, clients, &names, withTrafficSpec, withInternalVisibility)
 	if err != nil {
 		t.Fatalf("Failed to create initial Service: %v: %v", names.Service, err)
 	}
 
-	// Ensure that it only has cluster local addresses
-	for _, traffic := range resources.Route.Status.Traffic {
-		if !strings.HasSuffix(traffic.TrafficTarget.URL.Host, network.GetClusterDomainName()) {
-			t.Fatalf("Expected all subroutes to be cluster local")
+	for _, tag := range []string{subrouteTag1, subrouteTag2} {
+		if isClusterLocal, err := trafficWithTagIsClusterLocal(resources.Route.Status.Traffic, tag); err != nil {
+			t.Fatalf(err.Error())
+		} else if !isClusterLocal {
+			t.Fatalf("Expected subroute with tag %s to be cluster local", tag)
 		}
 	}
 
-	serviceName := fmt.Sprintf("%s-%s", tag, resources.Route.Name)
+	if !isRouteClusterLocal(resources.Route.Status) {
+		t.Fatalf("Expected route to be cluster local")
+	}
+	// change subroute 1 - private
+	serviceName := fmt.Sprintf("%s-%s", subrouteTag1, resources.Route.Name)
 	svc, err := clients.KubeClient.Kube.CoreV1().Services(test.ServingNamespace).Get(serviceName, metav1.GetOptions{})
 	if err != nil {
 		t.Fatalf("Failed to get k8s service to modify: %s", err.Error())
 	}
 
 	svcCopy := svc.DeepCopy()
-	labels.SetVisibility(&svcCopy.ObjectMeta, false)
+	labels.SetVisibility(&svcCopy.ObjectMeta, true)
 
-	patch, err := duck.CreatePatch(svc, svcCopy)
+	svcpatchBytes, err := test.CreateBytePatch(svc, svcCopy)
 	if err != nil {
-		t.Fatalf("Failed to create patch: %s", err.Error())
+		t.Fatalf(err.Error())
 	}
 
-	encPatch, err := patch.MarshalJSON()
+	_, err = clients.KubeClient.Kube.CoreV1().Services(test.ServingNamespace).Patch(serviceName, types.JSONPatchType, svcpatchBytes)
 	if err != nil {
-		t.Fatalf("Failed to marshal patch: %s", err.Error())
+		t.Fatalf("Failed to patch service: %s", err.Error())
+	}
+
+	// TODO: Add a better check here. ex. Wait 2s checking every 100ms to ensure it doesn't get the wrong visibility
+	time.Sleep(2 * time.Second) // TODO: Remove this after better check ^^
+	//if err = v1a1test.WaitForRouteState(clients.ServingAlphaClient, resources.Route.Name, v1a1test.IsRouteReady, "Route is ready"); err != nil {
+	//	t.Fatalf("Route did not become ready: %s", err.Error())
+	//}
+
+	// check route and all sub-routes are private
+	for _, tag := range []string{subrouteTag1, subrouteTag2} {
+		if isClusterLocal, err := trafficWithTagIsClusterLocal(resources.Route.Status.Traffic, tag); err != nil {
+			t.Fatalf(err.Error())
+		} else if !isClusterLocal {
+			t.Fatalf("Expected subroute with tag %s to be cluster local", tag)
+		}
+	}
+
+	if !isRouteClusterLocal(resources.Route.Status) {
+		t.Fatalf("Expected route to be cluster local")
 	}
-	_, err = clients.KubeClient.Kube.CoreV1().Services(test.ServingNamespace).Patch(serviceName, types.JSONPatchType, encPatch)
+
+	// change route - public
```

Comment:
is this comment for changing route to public?

Reply:
In this case, it's owned by the ksvc. We updated the ksvc's label instead.